%\VignetteIndexEntry{OmicAnalyzer User's Guide}
%\VignetteEncoding{UTF-8}

\documentclass{article}
\usepackage{amsmath}
\usepackage[breaklinks]{hyperref}
\title{OmicAnalyzer User's Guide}
\author{John Blischak}
\date{\today{} \hfill OmicAnalyzer \Sexpr{packageVersion("OmicAnalyzer")}}

\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle
\tableofcontents

<<echo=FALSE>>=
options(prompt = " ")
@

\section{Overview}

The goal of the OmicAnalyzer software is to facilitate the interactive
exploration of the results from an omics experiment. Using OmicAnalyzer, any
bioinformatician familar with the basics of the R programming language can
create and share a high-quality, comprehensive dashboard for interactive
investigation of the patterns and signals in the data.

The steps include:

\begin{enumerate}

  \item The bioinformatician analyzes the data from the omics experiment. This
  can be done in R or any other platform (e.g.
  \href{https://omicsoftdocs.github.io/ArraySuiteDoc/tutorials/ArrayStudio/ArrayStudio/}{Array
  Studio}).

  \item The bioinformatician registers the results using the OmicAnalyzer R
  functions. If the results were produced outside of R, they will need to be
  exported to files, and then imported into R.

  \item The bioinformatician uses OmicAnalyzer to export the analysis results to
  a study pacakge and install it.

  \item The bioinformatician can run the app locally or deploy it on a server.

  \item If the app is deployed, collaborators can access the dashboard directly
  in their web browser.

\end{enumerate}

\section{Structure of analysis results}

Before you start registering your data with OmicAnalyzer, it will be helpful to
have a high-level understanding of how OmicAnalyzer defines the main components
of the analysis results and how they relate to each other.

\subsection{A study and its models}

The largest unit of organization is the study. This corresponds to all the
experiments performed and analyses conducted in order to address a scientific
question. A study should contain all the relevant results that someone would
need to evaluate the scientific question. In more practical terms, any
analyses that uses overlapping biological samples should be included in the
same study (e.g. if you measured transcript and protein levels in the same
samples).

A study has one or models. The models can be very different (one model for
transcript levels and another for protein levels) or very similar (same exact
input data but differential expression performed with two different software
packages).

\includegraphics{images/models}

\subsection{Differential expression results}

One of the primary type of results that OmicAnalyzer displays for interactive
investigation is the statistical results from a differential expression
analysis. Here ``differential expression'' is broadly defined,
e.g. this could be differential methylation or any other molecular phenotype.

Each model has one or more tests. Each test (also commonly referred to as a
\href{https://en.wikipedia.org/wiki/Contrast_(statistics)}{contrast}) refers
to the statistical results from testing a single coefficient (or a combination
of coefficients) from the model.

For example, a model that compares cases versus controls will only have one
test:

\includegraphics{images/results-1}

A model that compares a control versus two different treatments will likely have
two tests:

\includegraphics{images/results-2}

And a model that compares three distinct groups will likely have 3 tests
(since $\binom{3}{2} = 3$ pairwise comparisons):

\includegraphics{images/results-3}

\subsection{Enrichments analysis}

A typical systems biology analysis to perform after a differential expression
analysis is to test for enrichments of the differentially expressed features
in known gene sets (e.g. \href{https://www.genome.jp/kegg/}{KEGG},
\href{https://reactome.org/}{Reactome}). The OmicAnalyzer app provides a
table, network, and various other visualizations to explore the output of
enrichment analyses.

Similar to the differential expression results, enrichments analyses are also
added per model and per test. However, there is also the additional category of
the annotation database that was used for the enrichment (e.g. KEGG, Reactome).
Thus each enrichments table corresponds to a given study-annotation-test
combination.

For example, imagine that enrichment analyses were performed with the KEGG and
Reactome annotations. A model that compares cases versus controls will have
two enrichments tables:

\includegraphics{images/enrichments-1}

A model that compares a control versus two different treatments will have
four enrichment tables, 2 for each of the tests:

\includegraphics{images/enrichments-2}

And a model that compares three distinct groups will have 6 enrichment tables,
2 for each of the 3 tests:

\includegraphics{images/enrichments-3}

\subsection{Additional information}

The primary data sets are the differential expression results and the enrichments
results. However, the more information you supply about the experimental data,
the more details will be included in the app.

For example, you can include metadata about the features, metadata about the
samples, the individual measurements (expression, methylation,
phosphorylation, etc.).

This additional metadata can be added per model (e.g. different feature
metadata for RNA versus protein measurements) or shared across models (e.g. if
the same samples are used in each model).

\section{Step-by-step example}

<<>>=
library(OmicAnalyzer)
@

<<>>=
data("RNAseq123")
ls()
@

\end{document}
