%\VignetteIndexEntry{OmicAnalyzer User's Guide}
%\VignetteEncoding{UTF-8}

\documentclass[12pt, letterpaper]{article}

\usepackage{fancyvrb}
\usepackage{amsmath}
\usepackage{float}
\usepackage[textwidth=6in]{geometry}
\usepackage[colorlinks=true,urlcolor=blue,breaklinks]{hyperref}
\usepackage[utf8]{inputenc}

\usepackage{color}
\definecolor{blue}{rgb}{0,0,0.5}

\usepackage{Sweave}
\RecustomVerbatimEnvironment{Sinput}{Verbatim}{
  fontshape=n,
  formatcom=\color{blue}
}

\title{OmicAnalyzer User's Guide}
\author{John Blischak}
\date{\today{} \hfill OmicAnalyzer \Sexpr{packageVersion("OmicAnalyzer")}}

\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle
\tableofcontents

<<setup, echo=FALSE>>=
if (!interactive()) options(prompt = " ", continue = " ", width = 70)
@

\section{Overview}

The goal of the OmicAnalyzer software is to facilitate the interactive
exploration of the results from an omics experiment. Using OmicAnalyzer, any
bioinformatician familar with the basics of the R programming language can
create and share a high-quality, comprehensive dashboard for interactive
investigation of the patterns and signals in the data.

The steps include:

\begin{enumerate}

  \item The bioinformatician analyzes the data from the omics experiment. This
  can be done in R or any other platform (e.g.
  \href{https://omicsoftdocs.github.io/ArraySuiteDoc/tutorials/ArrayStudio/ArrayStudio/}{Array
  Studio}).

  \item The bioinformatician registers the results using the OmicAnalyzer R
  functions. If the results were produced outside of R, they will need to be
  exported to files, and then imported into R.

  \item The bioinformatician uses OmicAnalyzer to export the analysis results to
  a study pacakge and install it.

  \item The bioinformatician can run the app locally or deploy it on a server.

  \item If the app is deployed, collaborators can access the dashboard directly
  in their web browser.

\end{enumerate}

\section{Structure of analysis results}

Before you start registering your data with OmicAnalyzer, it will be helpful to
have a high-level understanding of how OmicAnalyzer defines the main components
of the analysis results and how they relate to each other.

\subsection{A study and its models}

The largest unit of organization is the study. This corresponds to all the
experiments performed and analyses conducted in order to address a scientific
question. A study should contain all the relevant results that someone would
need to evaluate the scientific question. In more practical terms, any
analyses that uses overlapping biological samples should be included in the
same study (e.g. if you measured transcript and protein levels in the same
samples).

A study has one or models. The models can be very different (one model for
transcript levels and another for protein levels) or very similar (same exact
input data but differential expression performed with two different software
packages).

\begin{figure}[H]
  \includegraphics{images/models}
  \centering
  \label{fig:models}
\end{figure}

\subsection{Differential expression results}

One of the primary type of results that OmicAnalyzer displays for interactive
investigation is the statistical results from a differential expression
analysis. Here ``differential expression'' is broadly defined,
e.g. this could be differential methylation or any other molecular phenotype.

Each model has one or more tests. Each test (also commonly referred to as a
\href{https://en.wikipedia.org/wiki/Contrast_(statistics)}{contrast}) refers
to the statistical results from testing a single coefficient (or a combination
of coefficients) from the model.

For example, a model that compares cases versus controls will only have one
test:

\begin{figure}[H]
  \includegraphics{images/results-1}
  \centering
  \label{fig:results-1}
\end{figure}

A model that compares a control versus two different treatments will likely have
two tests:

\begin{figure}[H]
  \includegraphics{images/results-2}
  \centering
  \label{fig:results-2}
\end{figure}

And a model that compares three distinct groups will likely have 3 tests
(since $\binom{3}{2} = 3$ pairwise comparisons):

\begin{figure}[H]
  \includegraphics{images/results-3}
  \centering
  \label{fig:results-3}
\end{figure}

\subsection{Enrichment analysis}

A typical systems biology analysis to perform after a differential expression
analysis is to test for enrichment of the differentially expressed features
in terms from curated annotation databases (e.g. \href{https://www.genome.jp/kegg/}{KEGG},
\href{https://reactome.org/}{Reactome}). The OmicAnalyzer app provides a
table, network, and various other visualizations to explore the output of
enrichment analyses.

Similar to the differential expression results, enrichment analyses are also
added per model and per test. However, there is also the additional category of
the annotation database that was used for the enrichment (e.g. KEGG, Reactome).
Thus each enrichments table corresponds to a given study-annotation-test
combination.

For example, imagine that enrichment analyses were performed with the KEGG and
Reactome annotations. A model that compares cases versus controls will have
two enrichments tables:

\begin{figure}[H]
  \includegraphics{images/enrichments-1}
  \centering
  \label{fig:enrichments-1}
\end{figure}

A model that compares a control versus two different treatments will have
four enrichment tables, 2 for each of the tests:

\begin{figure}[H]
  \includegraphics{images/enrichments-2}
  \centering
  \label{fig:enrichments-2}
\end{figure}

And a model that compares three distinct groups will have 6 enrichment tables,
2 for each of the 3 tests:

\begin{figure}[H]
  \includegraphics{images/enrichments-3}
  \centering
  \label{fig:enrichments-3}
\end{figure}

\subsection{Additional information}

The primary data sets are the differential expression results and the enrichments
results. However, the more information you supply about the experimental data,
the more details will be included in the app.

For example, you can include metadata about the features, metadata about the
samples, or the individual assay measurements (expression, methylation,
phosphorylation, etc.).

This additional metadata can be added per model (e.g. different feature
metadata for RNA versus protein measurements) or shared across models (e.g. if
the same samples are used in each model).

\section{Step-by-step example}

Now it's time to convert your results into an OmicAnalyzer study! Below you will
see how to:

\begin{enumerate}

  \item Create a new OmicAnalyzer study

  \item Add your existing results to your OmicAnalyzer study

  \item Install your OmicAnalyzer study as an R package

  \item Start the app to interactively explore your results

\end{enumerate}

\subsection{Example data: RNAseq123}

As an illustrative example, this vignette uses the results from the
Bioconductor workflow package
\href{https://bioconductor.org/packages/release/workflows/html/RNAseq123.html}{RNAseq123}
\cite{RNAseq123}. It is a limma+voom analysis of RNA-seq data from 3 cell
populations in the mouse mammary gland collected by Sheridan et al., 2015
\cite{Sheridan2015}.

The 3 cell populations are basal, luminal progenitor (LP), and mature luminal (ML).
In the RNAseq123 analysis, the primary differential expression results obtained
are for the tests Basal versus LP and Basal versus ML. This is similar to
the design diagrammed in Figure \ref{fig:results-2}.

\begin{figure}[H]
  \includegraphics{images/results-rna}
  \centering
  \label{fig:results-rna}
\end{figure}

Furthermore they performed enrichment analysis with one annotation database:
the Broad Institute's MSigDB c2 collection, which was converted converted
from human to mouse gene identifiers by
\href{http://bioinf.wehi.edu.au/software/MSigDB/}{WEHI Bioinformatics}.

\begin{figure}[H]
  \includegraphics{images/enrichments-rna}
  \centering
  \label{fig:enrichchments-rna}
\end{figure}

To obtain their results, we ran their script
\href{https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.R}{limmaWorkflow.R}
and saved a subset of the objects to use in this vignette.

<<data>>=
data("RNAseq123", package = "OmicAnalyzer")
ls()
@

You are welcome to follow along with these example data if you like. However,
the main goal of this document is for you to input your own study data into
OmicAnalyzer. The particular details of this specific study are not important.
It happens to be an RNA-seq analysis of mice that was analyzed with limma+voom,
but none of that has to apply to your own study.

\subsection{Create an OmicAnalyzer study}

The first step is to load the OmicAnalyzer package into the current R session.

<<package>>=
library(OmicAnalyzer)
@

Next run \texttt{createStudy()} to create your OmicAnalyzer study object. The
first argument is the name of the study. The second argument (optional) is a description of
the study.

<<createStudy>>=
study <- createStudy("vignetteExample",
                     "Bioc workflow package converted to OmicAnalyzer")
@

Because the \texttt{name} will be used when naming the study package, it must
follow these rules that apply to all R package names:

\begin{enumerate}

  \item Begin with a letter

  \item End with a letter or a number

  \item Be at least two characters long

  \item Only contain alphanumeric characters and periods (full stops)

\end{enumerate}

\subsection{Add differential expression results}

<<>>=
head(basal.vs.lp)
@

<<>>=
head(cam.BasalvsLP)
@

\phantomsection
\addcontentsline{toc}{section}{References}
\begin{thebibliography}{2}

  \bibitem{RNAseq123}
    Law CW, Alhamdoosh M, Su S, Dong X, Tian L, Smyth GK, Ritchie ME.
    \href{https://f1000research.com/articles/5-1408/v3}
    {RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR [version 3; peer review: 3 approved].}
    F1000Research 2018, 5:1408

  \bibitem{Sheridan2015}
     Sheridan, J.M., Ritchie, M.E., Best, S.A. et al.
     \href{https://doi.org/10.1186/s12885-015-1187-z}
     {A pooled shRNA screen for regulators of primary mammary stem and progenitor cells identifies roles for \textit{Asap1} and \textit{Prox1}.}
     BMC Cancer 15, 221 (2015).

\end{thebibliography}

\end{document}
